proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,222 @@
+epoch rewards asynchronously. Instead of distributing rewards at the start of
+the epoch synchronously, we propose an asynchronous distribution method that
+spreads the rewards payout over multiple blocks.
+
+
+## Detailed Design
+
+The major bottleneck is paying out rewards to stake accounts, which is about
+550K, while rewards for vote accounts are around 1.5K. Given the relatively
+small number of vote rewards, it make sense to keep vote rewards payout the
+same. They can still be paid out synchronously and efficiently at the first
+block of the epoch boundary. This reduce the impact on rewards for vote account
+and also simplify the over changes, and let us focus on solving the primary
+bottleneck - Stake Rewards. Stake rewards are going to be paid out
+asynchronously over multiple blocks.
+
+For async stake rewards payout, we will separate the computation of rewards
+from the actual distribution of rewards at the epoch boundary by dividing the
+process into two distinct phases: (1) rewards calculation, and (2) rewards
+distribution.
+
+
+### Rewards Calculation
+
+In contrast to the current method of computing epoch rewards, which involves
+directly computing rewards at the first slot of the epoch boundary in the bank
+runtime, the new approach spreads the reward computation over multiple blocks.
+
+As an example, in the Solana Labs validator client, the reward computation is
+moved to a background service known as the `EpochRewardCalculationService`.
+This service starts automatically when a validator starts and listens to a
+designated channel for any incoming rewards calculation requests, performing
+the necessary rewards calculations accordingly.
+
+When the block on the network crosses the epoch boundary, a reward computation
+request based on stake delegation and vote states of vote accounts at the epoch
+boundary is sent to the `EpochRewardCalculationService` to start the process of
+epoch reward computation. This marks the start of the reward computation phase.
+
+The reward computation phase lasts for `N` blocks since the beginning of the
+epoch boundary. When reaching block height `N` after the start of the
+`reward calculation phase`, the bank will mark it the end the `reward calculation phase`.
+A conservative value of 1,000 is chosen for `N` to accommodate the most of the
+validators on the network.
+
+At this point, the reward computation result should be available. In case that
+the reward computation result is not available yet at the `N_th_` block-height,
+the bank runtime will block and wait until the computation is finished, before
+processing any blocks with height greater than `N`.
+
+At the end of the reward computation phase (end of block height `N`), a set of
+sysvar accounts will be updated - `EpochRewardProof` and `RewardReserve`.
+
+`EpochRewardProof` sysvar account maintains a history of epoch rewards. The
+internal format of `EpochRewardProof` is an array of entries,
+```
+[(epoch, (reward_amount, root_hash))]
```

Comment:
What is the `root_hash`, and how is it computed and used?

`reward_amount` is `total_reward_amount` correct?

Reply:
It would be useful for to include the number of `EpochRewardReserve` accounts to expect for each epoch